pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,399 @@
+	klog.V(2).InfoS("Uploading support bundle collection", "name", supportBundle.Name)
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("support bundle upload to file server failed during setting offset: %v", err)
+
+	}
+	// fileServer.URL should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path
+	var url string
+	if strings.Contains(supportBundle.FileServer.URL, "://") {
+		arr := strings.Split(supportBundle.FileServer.URL, "://")
+		if strings.ToLower(arr[0]) != string(sftpProtocol) {
+			return fmt.Errorf("only sftp is supported for now, protocol %s is not supported", strings.ToUpper(arr[0]))
+		}
+		url = supportBundle.FileServer.URL[7:]
+	} else {
+		url = supportBundle.FileServer.URL
+	}
+
+	triesLeft := uploadToFileServerTries
+	var uploadErr error
+	for triesLeft > 0 {
+		if uploadErr = c.uploadToFileServer(supportBundle.Name, url, &supportBundle.Authentication, outputFile); uploadErr == nil {
+			return nil
+		}
+		triesLeft -= 1
+		if triesLeft == 0 {
+			return fmt.Errorf("aborting upload support bundle after %d attempts", uploadToFileServerTries)
+		}
+		klog.InfoS("Failed to upload support bundle", "UploadError", uploadErr, "TriesLeft", triesLeft)
+		time.Sleep(uploadToFileServerRetryDelay)
+	}
+	return nil
+}
+
+func (c *SupportBundleController) uploadToFileServer(bundleName string, url string, serverAuth *cpv1b2.BundleServerAuthConfiguration, tarGzFile io.Reader) error {
+	strArr := strings.Split(url, "/")
+	address := strArr[0]
+	targetPath := url[len(address):]
+	path := path.Join(targetPath, c.nodeName+"_"+bundleName+".tar.gz")
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// #nosec G106: skip host key check here and users can specify their own checks if needed
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	}
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return err
+	}
+	return uploader.upload(address, path, cfg, tarGzFile)
+}
+
+func (c *SupportBundleController) getUploaderByProtocol(protocol ProtocolType) (uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+type uploader interface {
+	upload(addr string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error
+}
+
+type sftpUploader struct {
+}
+
+func (uploader *sftpUploader) upload(address string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error {
+	conn, err := ssh.Dial("tcp", address, config)
+	if err != nil {
+		return fmt.Errorf("error when connecting to fs server: %w", err)
+	}
+	sftpClient, err := sftp.NewClient(conn)
+	if err != nil {
+		return fmt.Errorf("error when setting up sftp client: %w", err)
+	}
+	defer func() {
+		if err := sftpClient.Close(); err != nil {
+			klog.Errorf("Error when closing sftp client: %v", err)
```

Comment:
should use structured logging. It applies to several new error logs.

Reply:
done